bokeh/models/sources.py

```diff
@@ -208,6 +201,38 @@ def _data_from_groupby(group):
         '''
         return ColumnDataSource._data_from_df(group.describe())
 
+    @staticmethod
+    def _df_index_name(df):
+        ''' Return the Bokeh-appropriate column name for a DataFrame index
+
+        If there is no named index, then `"index" is returned.
+
+        If there is a single named index, then ``df.index.name`` is returned.
+
+        If there is a multi-index, and the index names are all strings, then
+        the names are joined with '_' and the result is returned, e.g. for a
+        multi-index ``['ind1', 'ind2']`` the result will be "ind1_ind2".
+        Otherwise if any index name is not a string, the fallback name "index"
+        is returned.
+
+        Args:
+            df (DataFrame) : the DataFrame to find an index name for
+
+        Returns:
+            str
+
+        '''
+        if df.index.name:
+            return df.index.name
+        elif df.index.names:
+            try:
+                return "_".join(df.index.names)
+            except TypeError:
```

Comment:
Under what circumstances this happens?

Reply:
I call it "unnamed multi index" not sure if there is some other proper name for it. There's a test for it in any case: https://github.com/bokeh/bokeh/pull/7379/files#diff-16c9ca614fbc1d71df3e42fa8df21265R187

Reply:
<img width="771" alt="screen shot 2018-01-08 at 09 32 53" src="https://user-images.githubusercontent.com/1078448/34677809-eda67012-f456-11e7-9f1b-6ee3571a7185.png">

Reply:
I see. OK.